File: app/src/services/monitor.ts
/**
 * 监控命令组封装
 *
 * 对接后端 `monitor_*` commands 与 `monitor_metrics` 事件。
 */

import { invoke, listen } from './invoke'
import type { MonitorSample, SessionHealth } from '@/types/monitor'

/** 启动某会话的监控采样（intervalMs 默认 3000） */
export function start(sessionId: string, intervalMs = 3000): Promise<void> {
  return invoke<void>('monitor_start', { sessionId, intervalMs })
}

/** 停止某会话的监控采样 */
export function stop(sessionId: string): Promise<void> {
  return invoke<void>('monitor_stop', { sessionId })
}

/** 当前采样中的会话列表 */
export function list(): Promise<string[]> {
  return invoke<string[]>('monitor_list')
}

/** 所有会话的健康/流量信息（会话健康面板轮询） */
export function sessionStatsAll(): Promise<SessionHealth[]> {
  return invoke<SessionHealth[]>('session_stats_all')
}

/** 订阅 monitor_metrics 事件（返回 unlisten） */
export function onMetrics(cb: (sample: MonitorSample) => void): Promise<() => void> {
  return listen<MonitorSample>('monitor_metrics', cb)
}

File: crates/desktop/src/commands/monitor.rs
//! # 监控命令组
//!
//! 启停按会话的指标采样（数据经 `monitor_metrics` 事件推送）。

use tauri::{AppHandle, State};

use crate::AppState;

/// 启动某会话的监控采样
///
/// 前端在 SSH 会话建立（或激活监控）后调用；
/// 数据通过 `monitor_metrics` 事件按 `interval_ms` 周期推送。
#[tauri::command]
pub async fn monitor_start(
    app: AppHandle,
    state: State<'_, AppState>,
    session_id: String,
    interval_ms: Option<u64>,
) -> Result<(), crate::error::AppError> {
    // 会话必须存在（未连接时报错给前端）
    if state.ssh_manager.get(&session_id).await.is_none() {
        return Err(crate::error::AppError::monitor(format!("会话 {session_id} 不存在或已断开")));
    }
    state
        .monitor_sampler
        .start(app, state.ssh_manager.clone(), session_id, interval_ms.unwrap_or(3000))
        .await;
    Ok(())
}

/// 停止某会话的监控采样
#[tauri::command]
pub async fn monitor_stop(
    state: State<'_, AppState>,
    session_id: String,
) -> Result<(), crate::error::AppError> {
    state.monitor_sampler.stop(&session_id).await;
    Ok(())
}

/// 当前采样中的会话列表
#[tauri::command]
pub async fn monitor_list(state: State<'_, AppState>) -> Result<Vec<String>, crate::error::AppError> {
    Ok(state.monitor_sampler.list().await)
}
